### DISCUSSION TITLE
The rebuilding nodes in the electron Node-RED

### DISCUSSION THREAD

**Discussion by kazuhitoyokoi** [node-red/designs#50] [Design Proposals]
As we talked about in [the pull request](https://github.com/node-red/designs/pull/30#issuecomment-747274009), the electron version of Node-RED needs to rebuild nodes that have version-specific binary because the binary is incompatible between Node.js in the Electron and normal Node.js which is installed in the local machine. In this discussion, I would like to clarify the timing when the rebuilding process is needed. In my understanding, there are two cases as follows.

- **Case 1: Rebuilding when changing the Node-RED environment**
 When a user changes the environment from normal Node-RED to Election Node-RED, there may be some existing nodes that the user installed using the normal Node-RED environment. In this case, the Electron version of Node-RED needs to rebuild all of the existing nodes before starting Node-RED.
 <img src="https://user-images.githubusercontent.com/20310935/116839941-10b95380-ac0f-11eb-91ce-b030cb8f90ce.png" width=300>

- **Case 2: Rebuilding after installing a new node**
 When a user installs a new node in the Electron Node-RED environment, the installed node requires rebuilding for the electron environment.
 <img src="https://user-images.githubusercontent.com/20310935/116840014-51b16800-ac0f-11eb-8c97-d49f90719ffd.png" width=300>

The pre/postInstall hook is for the case 2 only. Therefore, we need to consider case 1.

**Comment by dceejay** [comment]
For the whole electron app - is it really using the existing ~/.node-red directory ? Are we really expecting users to be able to swap between electron use and an existing install ? How would they swap back ?

**Reply by kazuhitoyokoi** [reply]
The current implementation is using the `~/.node-red` directory. If users want to swap back to normal Node-RED, they need to execute `npm rebuild`.

As you asked, another directory like `~/.node-red-installer` may be better for the Electron app to separate it from the normal Node-RED environment.

**Reply by knolleary** [reply]
Right, this is all part of the basic standalone design that still hasn't been agreed. We need to agree the basic model of what the installer provides.
